src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Utility.cs

```diff
@@ -94,613 +71,275 @@ private static bool FirstCharInUInt32IsAscii(uint value)
-            // bits which are less significant than those that the addition would've acted on.
+                    offset = 0;
+                    found = first;
+                    goto Found256;
+                }
+                if (Test256(last))
+                {
+                    offset = bufferLength - 32;
+                    found = last;
+                    goto Found256;
+                }
 
-            // If there is fewer than one vector length remaining, skip the next aligned read.
+                return bufferLength;
 
-            if ((bufferLength & SizeOfVector128) == 0)
-            {
-                goto DoFinalUnalignedVectorRead;
+            Found256:
+                found = Vector256.GreaterThanOrEqual(found, Vector256.Create((byte)0x80));
+                return ComputeIndex(found) + offset;
             }
 
-            // At least one full vector's worth of data remains, so we can safely read it.
-            // Remember, at this point pBuffer is still aligned.
+            Vector512<byte> found512;
+            ref byte pCurrent = ref pBuffer;
+            ref byte pLast = ref Unsafe.Add(ref pBuffer, bufferLength - 64);
 
-            if (Sse2.IsSupported)
+            // SAFETY: We only use the value of .AsPointer to calculate the byte offset
+            // to the next 64B boundary. Should GC relocate the buffer, we will lose the
+            // alignment which is an acceptable trade-off over pinning the buffer, once
+            // the callers of this method are updated to use byrefs.
+            nuint toAlign = (nuint)Unsafe.AsPointer(ref pCurrent) % 64;
+            if (toAlign != 0)
             {
-                currentSseMask = (uint)Sse2.MoveMask(Sse2.LoadAlignedVector128(pBuffer));
-                if (ContainsNonAsciiByte_Sse2(currentSseMask))
+                Vector512<byte> first512 = Vector512.LoadUnsafe(ref pCurrent);
+                if (Test512(first512))
                 {
-                    goto FoundNonAsciiDataInCurrentChunk;
+                    found512 = first512;
+                    goto Found512;
                 }
+
+                pCurrent = ref Unsafe.Add(ref pCurrent, toAlign);
             }
-            else if (AdvSimd.Arm64.IsSupported)
+
+            while (!Unsafe.IsAddressGreaterThan(ref pCurrent, ref pLast))
             {
-                currentAdvSimdIndex = (uint)GetIndexOfFirstNonAsciiByteInLane_AdvSimd(AdvSimd.LoadVector128(pBuffer), bitmask);
-                if (ContainsNonAsciiByte_AdvSimd(currentAdvSimdIndex))
+                Vector512<byte> current512 = Vector512.LoadUnsafe(ref pCurrent);
+                if (!Test512(current512))
                 {
-                    goto FoundNonAsciiDataInCurrentChunk;
+                    // Better loop ordering
+                    pCurrent = ref Unsafe.Add(ref pCurrent, 64);
+                    continue;
                 }
+
+                found512 = current512;
+                goto Found512;
```

Comment:
We're trying to reduce the total number of `goto` cases and similar unsafe code as it complicates control flow analysis, general optimization work, and is easy to get wrong.

We should try and write this without instead, if possible.

Reply:
Calculating index, particularly on ARM64, takes a bit of space. Merging three potential places where it happens into a single label allows to save quite a bit of codegen. It's a simple "break out to end" from three places. At least I don't see a nicer way to write it...

Reply:
Use an aggressively inlined helper method instead?

Reply:
This goes back to producing much larger than ideal codegen - it's too complex for the JIT to realize it could do what is done here manually. I will try to make it work once again, but likely you'll see asm diffs here 🤷